packages/playground/blueprints/src/lib/compile.ts

```diff
@@ -116,33 +145,77 @@ export function compileBlueprint(
 			),
 		},
 		run: async (playground: UniversalPHP) => {
+			const progressMap = new Map<StepDefinition, Promise<void>>();
+			const progressPromises = new Map<StepDefinition, PromiseExecutor>();
+			for (const { step } of compiled) {
+				progressMap.set(
+					step,
+					new Promise((resolve, reject) => {
+						progressPromises.set(step, { resolve, reject });
+					})
+				);
+			}
+			onRunningSteps(progressMap);
+
 			try {
 				// Start resolving resources early
-				for (const { resources } of compiled) {
+				for (const { resources, step } of compiled) {
+					const stepPromise = progressPromises.get(
+						step
+					) as PromiseExecutor;
 					for (const resource of resources) {
 						resource.setPlayground(playground);
 						if (resource.isAsync) {
-							resource.resolve();
+							resource.resolve().catch(function (error) {
+								stepPromise.reject(error);
+							});
 						}
 					}
 				}
 
+				/**
+				 * When a compiled step's `run` method is called, it
+				 * uses `resolveArguments` to await dependent resources.
+				 */
 				for (const { run, step } of compiled) {
-					const result = await run(playground);
-					onStepCompleted(result, step);
+					const stepPromise = progressPromises.get(
+						step
+					) as PromiseExecutor;
+					try {
+						const result = await run(playground);
+						stepPromise.resolve(result);
+						onStepCompleted(result, step);
+					} catch (error) {
+						stepPromise.reject(error);
```

Comment:
we could also call these something else to explain more the use.
here's one random idea:

```js
const stepIndicator = stepIndicators.get( step );

…

stepIndicator.succeedWith( result );

…

stepIndicator.failWith( error );
```

or `succeededWith`/`failedWith`

```js
stepIndicators.set( step, { succeedWith: resolve, failWith: reject } );
```

Reply:
`stepPromise` has been renamed to `stepStatus`.